x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/EscapedFunctionsTests.java

```diff
@@ -0,0 +1,225 @@
+        Function f = (Function) parser.createExpression("POWER(?, {fn POWER({fn ABS(?)}, {fN ABS(?)})})",
+                asList(new SqlTypedParamValue(DataType.LONG, 1),
+                       new SqlTypedParamValue(DataType.LONG, 1),
+                       new SqlTypedParamValue(DataType.LONG, 1)));
+
+        assertEquals("POWER", f.functionName());
+        assertEquals(2, f.arguments().size());
+        Expression arg = f.arguments().get(1);
+        assertThat(arg, instanceOf(UnresolvedFunction.class));
+        UnresolvedFunction uf = (UnresolvedFunction) arg;
+        assertThat(uf.name(), is("POWER"));
+        assertEquals(2, uf.arguments().size());
+
+        List<Expression> args = uf.arguments();
+        arg = args.get(0);
+        assertThat(arg, instanceOf(UnresolvedFunction.class));
+        uf = (UnresolvedFunction) arg;
+        assertThat(uf.name(), is("ABS"));
+
+        arg = args.get(1);
+        assertThat(arg, instanceOf(UnresolvedFunction.class));
+        uf = (UnresolvedFunction) arg;
+        assertThat(uf.name(), is("ABS"));
+    }
+
+    public void testDateLiteral() {
+        Literal l = dateLiteral("2012-01-01");
+        assertThat(l.dataType(), is(DataType.DATE));
+    }
+
+    public void testDateLiteralValidation() {
+        ParsingException ex = expectThrows(ParsingException.class, () -> dateLiteral("2012-13-01"));
+        assertEquals("line 1:2: Invalid date received; Cannot parse \"2012-13-01\": Value 13 for monthOfYear must be in the range [1,12]",
+                ex.getMessage());
+    }
+
+    public void testTimeLiteralUnsupported() {
+        SqlIllegalArgumentException ex = expectThrows(SqlIllegalArgumentException.class, () -> timeLiteral("10:10:10"));
+        assertThat(ex.getMessage(), is("Time (only) literals are not supported; a date component is required as well"));
+    }
+
+    public void testTimeLiteralValidation() {
+        ParsingException ex = expectThrows(ParsingException.class, () -> timeLiteral("10:10:65"));
+        assertEquals("line 1:2: Invalid time received; Cannot parse \"10:10:65\": Value 65 for secondOfMinute must be in the range [0,59]",
+                ex.getMessage());
+    }
+
+    public void testTimestampLiteral() {
+        Literal l = timestampLiteral("2012-01-01 10:01:02.3456");
+        assertThat(l.dataType(), is(DataType.DATE));
+    }
+
+    public void testTimestampLiteralValidation() {
+        ParsingException ex = expectThrows(ParsingException.class, () -> timestampLiteral("2012-01-01T10:01:02.3456"));
+        assertEquals(
+                "line 1:2: Invalid timestamp received; Invalid format: \"2012-01-01T10:01:02.3456\" is malformed at \"T10:01:02.3456\"",
+                ex.getMessage());
+    }
+
+    public void testGUID() {
+        Literal l = guidLiteral("12345678-90ab-cdef-0123-456789abcdef");
+        assertThat(l.dataType(), is(DataType.KEYWORD));
+    }
+
+    public void testGUIDValidation() {
```

Comment:
Worth adding some more tests here? Like uppercase/lowercase combined letters and other letters from the alphabet?

Reply:
The upper/lower case mix is addressed in the following test. Nevertheless I've added a couple of tests that check the different validations of GUID.